team page: Display contributor count per repo.

eeshangarg opened the pull request:
Screenshot:
![total_contributors](https://user-images.githubusercontent.com/7251823/148840949-296a14e5-5995-4dae-9679-d6489a3996f1.png)

See [this CZO thread for more details](https://chat.zulip.org/#narrow/stream/2-general/topic/stats.20on.20.2Fteam.20page/near/1306415).

eeshangarg commented:
@alya Could you please take a look at this PR? Thanks!

@timabbott FYI :)

timabbott commented on templates/zerver/team.html:
```diff
@@ -141,6 +157,10 @@ <h1 id="our-amazing-community">Our amazing community</h1>
                 </script>
 
+                <script type="text/template" id="contributor-count-template">
+                    <p class="contributor-count">Total contributors: <%= contributor_count %></p>
```
Will this always be a grand total, or just the total for a given (sub)project?  It might be good to make that clear, e.g. "Total server contributors" or similar when you're in those tabs.

timabbott replied:
Or just "353 server contributors" might feel like cleaner language?

timabbott commented:
I posted some small feedback on the wording, and a request that we have the data for each data if that's doable.  Additionally, we should be sure to do some manual checking of the results; 1072 contributors is plausible, but I'd want to make sure we've done some double-checking before we claim it prominently on our website (I would have guessed between 900 and 1100 if asked).  One approach would be to (1) double-check the individual repository numbers against GitHub's presentation of that data, (2) randomly sample 10 GitHub usernames that are not on the zulip/zulip and manually confirm each is a contributor only to other repositories, (3) make sure we feel  the logic to avoid double-counting is correct.

eeshangarg commented:
@timabbott: I'll respond to your points of feedback one by one:

**Getting data for each repository**

As for getting the data for each repository, yes my PR already did that. I addressed your feedback around wording and also made it such that the count template links to each repo's contributor page. Here is a screenshot of the total contributor count from all repos:
![total_contributors](https://user-images.githubusercontent.com/7251823/148999835-bd4a66ed-146d-4534-b348-a3be225b1e7c.png)

Here is a screenshot of the contributor count from a tab for a specific repo, note that the count links to the repo's contributors page. I also used the repository's name here so that it is more explicit about which repository we are talking about, i.e. "zulip-mobile contributors" vs just "mobile contributors".
![mobile-contributors](https://user-images.githubusercontent.com/7251823/148999999-acd637d0-8280-4291-8973-1920093335db.png)

**Double checking individual repo numbers against GitHub's API representation**

There does seem to be a disconnect between the individual repository numbers versus what GitHub claims in their API responses. The numbers in the API responses are generally higher. For instance, go to https://github.com/zulip/zulip, you can scroll down and see the total contributor count on the right, it says 735. But the API response we get says 863. Here are the numbers for all the repositories:
  * zulip/zulip: 735 vs 863
  * zulip/zulip-desktop: 54 vs 60
  * zulip/zulip-mobile: 77 vs 84
  * zulip/python-zulip-api: 118 vs 144
  * zulip/zulip-js: 14 vs 14
  * zulip/zulipbot: 11 vs 12
  * zulip/zulip-terminal: 52 vs 55

Upon some deeper digging, I discovered a potential cause of this discrepancy. Turns out that the API response includes entries that have an email associated with a commit in our repository, but the associated GitHub account is no longer active. Here are two such examples in our `zulip-desktop` repo: [vbNETonIce](https://github.com/zulip/zulip/commits?author=vbnetonice@users.noreply.github.com) and [Jan Hrnko](https://github.com/zulip/zulip-desktop/commits?author=jan.hrnko@satoshilabs.com). Both have committed in `zulip-desktop` but you can't click on their GitHub profiles, because they probably don't exist anymore. The cause of the discrepancy is that the count displayed on the repository page does not take these commits from inactive GitHub profiles into account. However, I think that for all intents and purposes, we have every reason to believe that the API is returning accurate data here.

**Randomly sample 10 GitHub usernames that are not on the zulip/zulip**

We can't actually do this because [the contributors page for zulip/zulip](https://github.com/zulip/zulip/graphs/contributors) only shows the top 100 contributors, so there are tons of usernames that won't show up there but still had like 1 commit to `zulip/zulip`. However, in the raw data we get from the GitHub API, there are tons of usernames that contributed to other repositories but not to `zulip/zulip`.

**Make sure we feel the logic to avoid double-counting is correct**

Two points on this:
* Basic manual inspection of the raw data does not point to any duplicates.
* Our logic for preventing double-counting in `tools/fetch-contributor-data` (based on the data in`tools/duplicate_commits.json`) looks correct to me.

@timabbott That was a lot but some of the questions you asked warranted a deeper investigation. This PR is now ready for another review! Thanks! :)

timabbott commented:
Let's generate a list of the 7 extra users for zulip-mobile, which doesn't have the 100 limit problem, and figure out what the story is with them (I'm sure the mobile team folks can help).  That should get to the bottom of why these numbers are different.

andersk commented:
The extra horizontal line under the contributor count should be removed.

Suggest clarifying the wording and reducing the length of the link:

> [84 contributors](https://github.com/zulip/zulip-mobile/graphs/contributors) to zulip/zulip-mobile

Or if we don’t want to bother debugging the discrepancy between our numbers and GitHub’s numbers:

> 84 contributors to [zulip/zulip-mobile](https://github.com/zulip/zulip-mobile)

eeshangarg commented:
@andersk Thanks for the feedback! Yes, I am already in the process of debugging the contributor discrepancies. Updated screenshot after your suggested changes were implemented:
![api-contributors](https://user-images.githubusercontent.com/7251823/149214353-ad869dbd-421a-4e6a-929c-a7dc8adb9202.png)

eeshangarg commented:
@timabbott Just pushed a commit that excludes entries by `dependabot`. This PR is ready for another review! Thanks! :)

timabbott commented:
I like the idea of having the link be to the repository as Anders suggested above, rather than the GitHub contributors page, now that we know the GitHub page count is inaccurate (in excluding any contributors whose email address is not associated with a GitHub account).

timabbott commented on static/js/portico/team.js:
```diff
@@ -58,6 +58,10 @@ function get_display_name(contributor) {
 
+function exclude_github_username(contributor, excluded_username) {
+    return contributor.github_username !== excluded_username;
+}
```
Let's have the function be `exclude_bot_contributors` and have it exclude dependabot -- I think that's a more extensible design if we end up with other automation creating commits in the future.

timabbott commented on static/js/portico/team.js:
```diff
@@ -92,8 +98,11 @@ export default function render_tabs() {
+                const filtered = contributors_list
+                    .filter((c) => c[repo_name])
+                    // Exclude entries by dependabot
+                    .filter((c) => exclude_github_username(c, "dependabot[bot]"));
```
It's not great that we need to do this filtering in two different places, since that's the kind of pattern likely to lead to bugs in the future -- ideally we'd do it in a single place when sourcing the data, either in the tool that fetches data from GitHub or in a preprocessing step.

eeshangarg commented:
@timabbott I just updated this PR with the following changes:
- Addressed your comments on the PR above.
- Added a commit that displays the number of contributors with 100+ contributions for each repository.
- Added a ReadTheDocs page explaining how the contributors' stats are calculated.

This PR is now ready for review! Thanks!

eeshangarg commented:
Updated screenshots:
![api-contributors](https://user-images.githubusercontent.com/7251823/149597023-4147ed3a-c60d-4a15-90ce-628a38e4d39a.png)

![total_contributors](https://user-images.githubusercontent.com/7251823/149597038-855045c3-e69b-4782-9e11-868524a96b7a.png)

alya commented:
The parenthetical seems hard to parse as currently written. I would put it after the name of the repo.

alya commented:
For the documentation, I think the most important part might be having instructions for how to fix things if you change your GitHub email. @eeshangarg  Could you take a pass on adding that?

The page also feels a bit out of place to me, but I'm not sure where it should go otherwise.

eeshangarg commented:
@alya Thanks for the feedback! I addressed your feedback:

* Added the parenthetical after the name of the repository. Here is an updated screenshot:
![contributors](https://user-images.githubusercontent.com/7251823/150229071-2577beeb-30b2-4c55-89bb-4f1ad29e9670.png)

* Added a section on how to fix things if one changes their GitHub email. I actually reached out to GitHub about this issue a couple of months ago, so turns out I already had the required context to write this section. Essentially, unless you maintain access to the old email and add it back to your GitHub account, you can't really link your old contributions to your profile. GitHub requires you to confirm all emails added to your account.

* As for where the document should live, the reason I put it with the contribution docs is that this page would be more likely to be read by contributors, so it made sense to put it where a lot of our docs on contribution already live.

eeshangarg commented:
@timabbott This is ready for a review! Thanks! :)

timabbott commented:
I merged all but the last commit (the new ReadTheDocs page) as the series ending with f2e7b92b024cfec6a9babace8298de9537925347.  I'm going to attempt to do an edit pass on the remaining commit and push it back for further feedback.

alya commented:
> Added a section on how to fix things if one changes their GitHub email. I actually reached out to GitHub about this issue a couple of months ago, so turns out I already had the required context to write this section. Essentially, unless you maintain access to the old email and add it back to your GitHub account, you can't really link your old contributions to your profile. GitHub requires you to confirm all emails added to your account.

Hm, maybe I'm confused -- I was referring to [this comment on CZO](https://chat.zulip.org/#narrow/stream/2-general/topic/stats.20on.20.2Fteam.20page/near/1307517):

> https://github.com/zulip/zulip/blob/main/.mailmap
> If you want to correct your contributor data, you should submit a PR modifying that file to treat your accounts as the same person. (GitHub, like Git, uses that file to understand that).

eeshangarg commented:
@alya Ah, I forgot about that one. I am not sure if that necessarily helps us circumvent the issue I wrote about. I'm guessing GitHub still won't display them as your contributions on your profile if the associated email is not on your GitHub account, but it could be that I am missing something here. @timabbott is doing an edit pass on the doc soon, pinging him so he can add a section on this if need be. Thanks!

timabbott commented:
I think this page should absolutely talk about the `mailmap` file as Alya suggests.  I don't quite understand the uncertainty @eeshangarg -- is the point that you're not sure how the list of email addresses on your GitHub profile interact with the `.mailmap` file?  That seems like something we could test.

I also think we should retitle it to "Contributor statistics" -- that feels like a more appropriate framing of what this page is about, and copy in a paragraph from our release blog posts setting context that this only counts code contributions, and that we as a community value contributions that are not counted by these statistics.

eeshangarg commented:
@timabbott Yeah, I wasn't sure how `mailmap` helps in this case. Does `mailmap` change how GitHub behaves? Or does it just change how we present the contributions on our page? My guess is that it doesn't help with the former question.

It seems like the easiest way to link your past contributions from an email you no longer have access to is to just add the email address to the GitHub account but not verify it, so that should really be it in the sense that we'll probably display it correctly as long as GitHub aggregates the data correctly.

timabbott commented:
https://stackoverflow.com/questions/53629125/does-github-consider-mailmap-for-contribution-graph suggests that GitHub uses `mailmap` files.  I would expect it to basically group all commits for a given user into the GitHub account associated with the "after .mailmap translations" email address.

But yeah, it'd be reasonable to mention both options in the paragraph we add here.

alya commented:
@timabbott the remaining work on this PR is for you to review and finalize the contributor documentation update.

timabbott commented:
I did a basic writing pass on this so that we check this off as content-complete. @alya will likely do a follow-up one.
